feat: Improve @CascadingUpdateShadowVariable logic (#1020)
TimefoldAI/timefold-solver: docs/src/modules/ROOT/pages/using-timefold-solver/modeling-planning-problems.adoc

=== Updating tail chains

The annotation `@CascadingUpdateShadowVariable` provides a built-in listener that updates a set of connected elements.
Timefold Solver triggers a user-defined logic after all events are processed.
Hence, the related listener is the final one executed during the event lifecycle.
Moreover,

> it automatically propagates changes to the next elements when the value of the related shadow variable changes.

Reviewer: ```suggestion
it automatically propagates changes to the subsequent elements in the list when the value of the related shadow variable changes.
```

The `targetMethodName` refers to the user-defined logic that updates the annotated shadow variable.
The method must be implemented in the defining entity class, be non-static, and not include any parameters.

In the previous example,
the cascade update listener calls `updateArrivalTime` after all shadow variables have been updated,
including `vehicle` and `previousCustomer`.
It then automatically calls `updateArrivalTime` for the subsequent customers
and stops when the `arrivalTime` value does not change after running target method
or when it reaches the end.

[WARNING]
====
A user-defined logic can only change shadow variables.

> It must never change a genuine planning variable or a problem fact.

Reviewer: ```suggestion
Changing a genuine planning variable or a problem fact will result in score corruption.
```

====

[NOTE]
====
When distinct target methods are used by separate `@CascadingUpdateShadowVariable` variables in the same model,

> there will be no guarantees regarding which target method is executed first.

Reviewer: ```suggestion
the order of their execution is undefined.
```

It must never change a genuine planning variable or a problem fact.
====

> [WARNING]

Reviewer: ```suggestion
[NOTE]
```

when the source shadow variable is `@CascadingUpdateShadowVariable`.
====

[WARNING]
====

> The user's logic is responsible for defining the order in which each variable is updated.

Reviewer: I don't understand what this is trying to say. Maybe it needs a bit of context?

Not sure if it's a warning. If it has more text explaining why this is bad, then I say it could be a warning.

Author: I believe it caused more confusion and I will remove it.
